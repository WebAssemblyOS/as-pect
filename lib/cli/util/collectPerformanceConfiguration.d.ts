import { IYargs } from "./IYargs";
import { IPerformanceConfiguration } from "../../util/IPerformanceConfiguration";
/**
 * This method collects the performance configuration values byref.
 *
 * @param {IYargs} yargs - The command line arguments
 * @param {IPerformanceConfiguration} performanceConfiguration - The effective performance configuration.
 */
export declare function collectPerformanceConfiguration(yargs: IYargs, performanceConfiguration: IPerformanceConfiguration): void;
//# sourceMappingURL=collectPerformanceConfiguration.d.ts.map