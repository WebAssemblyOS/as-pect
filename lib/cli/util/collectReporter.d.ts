import { TestReporter } from "../../test/TestReporter";
import { IYargs } from "./IYargs";
/**
 * This method inspects the command line arguments and returns the corresponding TestReporter.
 *
 * @param {IYargs} yargs - The command line arguments.
 */
export declare function collectReporter(yargs: IYargs): TestReporter;
//# sourceMappingURL=collectReporter.d.ts.map