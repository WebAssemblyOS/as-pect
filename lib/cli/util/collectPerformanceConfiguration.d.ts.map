{"version":3,"file":"collectPerformanceConfiguration.d.ts","sourceRoot":"","sources":["../../../src/cli/util/collectPerformanceConfiguration.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,MAAM,EAAE,MAAM,UAAU,CAAC;AAClC,OAAO,EAAE,yBAAyB,EAAE,MAAM,sCAAsC,CAAC;AAEjF;;;;;GAKG;AACH,wBAAgB,+BAA+B,CAAC,KAAK,EAAE,MAAM,EAAE,wBAAwB,EAAE,yBAAyB,QAwBjH"}