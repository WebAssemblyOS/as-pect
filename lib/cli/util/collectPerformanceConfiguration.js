(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * This method collects the performance configuration values byref.
     *
     * @param {IYargs} yargs - The command line arguments
     * @param {IPerformanceConfiguration} performanceConfiguration - The effective performance configuration.
     */
    function collectPerformanceConfiguration(yargs, performanceConfiguration) {
        if (yargs.argv.hasOwnProperty("performance"))
            performanceConfiguration.enabled = yargs.argv.performance !== "false";
        // gather all the flags
        if (yargs.argv.hasOwnProperty("maxSamples"))
            performanceConfiguration.maxSamples = parseFloat(yargs.argv.maxSamples.toString());
        if (yargs.argv.hasOwnProperty("maxTestRunTime"))
            performanceConfiguration.maxTestRunTime = parseFloat(yargs.argv.maxTestRunTime.toString());
        if (yargs.argv.hasOwnProperty("maxTestRunTime"))
            performanceConfiguration.maxTestRunTime = parseFloat(yargs.argv.maxTestRunTime.toString());
        if (yargs.argv.hasOwnProperty("roundDecimalPlaces"))
            performanceConfiguration.roundDecimalPlaces = parseFloat(yargs.argv.roundDecimalPlaces.toString());
        if (yargs.argv.hasOwnProperty("reportMedian"))
            performanceConfiguration.reportMedian = yargs.argv.reportMedian !== "false";
        if (yargs.argv.hasOwnProperty("reportAverage"))
            performanceConfiguration.reportAverage = yargs.argv.reportAverage !== "false";
        if (yargs.argv.hasOwnProperty("reportStandardDeviation"))
            performanceConfiguration.reportStandardDeviation = yargs.argv.reportStandardDeviation !== "false";
        if (yargs.argv.hasOwnProperty("reportMax"))
            performanceConfiguration.reportMax = yargs.argv.reportMax !== "false";
        if (yargs.argv.hasOwnProperty("reportMin"))
            performanceConfiguration.reportMin = yargs.argv.reportMin !== "false";
        if (yargs.argv.hasOwnProperty("reportVariance"))
            performanceConfiguration.reportVariance = yargs.argv.reportVariance !== "false";
    }
    exports.collectPerformanceConfiguration = collectPerformanceConfiguration;
});
//# sourceMappingURL=collectPerformanceConfiguration.js.map