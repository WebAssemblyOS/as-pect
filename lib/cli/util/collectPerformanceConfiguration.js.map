{"version":3,"file":"collectPerformanceConfiguration.js","sourceRoot":"","sources":["../../../src/cli/util/collectPerformanceConfiguration.ts"],"names":[],"mappings":";;;;;;;;;;;IAGA;;;;;OAKG;IACH,SAAgB,+BAA+B,CAAC,KAAa,EAAE,wBAAmD;QAChH,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,aAAa,CAAC;YAC1C,wBAAwB,CAAC,OAAO,GAAG,KAAK,CAAC,IAAI,CAAC,WAAW,KAAK,OAAO,CAAC;QACxE,uBAAuB;QACvB,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,YAAY,CAAC;YACzC,wBAAwB,CAAC,UAAU,GAAG,UAAU,CAAC,KAAK,CAAC,IAAI,CAAC,UAAU,CAAC,QAAQ,EAAE,CAAC,CAAC;QACrF,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,gBAAgB,CAAC;YAC7C,wBAAwB,CAAC,cAAc,GAAG,UAAU,CAAC,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,QAAQ,EAAE,CAAC,CAAC;QAC7F,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,gBAAgB,CAAC;YAC7C,wBAAwB,CAAC,cAAc,GAAG,UAAU,CAAC,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,QAAQ,EAAE,CAAC,CAAC;QAC7F,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,oBAAoB,CAAC;YACjD,wBAAwB,CAAC,kBAAkB,GAAG,UAAU,CAAC,KAAK,CAAC,IAAI,CAAC,kBAAkB,CAAC,QAAQ,EAAE,CAAC,CAAC;QACrG,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,cAAc,CAAC;YAC3C,wBAAwB,CAAC,YAAY,GAAG,KAAK,CAAC,IAAI,CAAC,YAAY,KAAK,OAAO,CAAC;QAC9E,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,eAAe,CAAC;YAC5C,wBAAwB,CAAC,aAAa,GAAG,KAAK,CAAC,IAAI,CAAC,aAAa,KAAK,OAAO,CAAC;QAChF,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,yBAAyB,CAAC;YACtD,wBAAwB,CAAC,uBAAuB,GAAG,KAAK,CAAC,IAAI,CAAC,uBAAuB,KAAK,OAAO,CAAC;QACpG,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,WAAW,CAAC;YACxC,wBAAwB,CAAC,SAAS,GAAG,KAAK,CAAC,IAAI,CAAC,SAAS,KAAK,OAAO,CAAC;QACxE,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,WAAW,CAAC;YACxC,wBAAwB,CAAC,SAAS,GAAG,KAAK,CAAC,IAAI,CAAC,SAAS,KAAK,OAAO,CAAC;QACxE,IAAI,KAAK,CAAC,IAAI,CAAC,cAAc,CAAC,gBAAgB,CAAC;YAC7C,wBAAwB,CAAC,cAAc,GAAG,KAAK,CAAC,IAAI,CAAC,cAAc,KAAK,OAAO,CAAC;IACpF,CAAC;IAxBD,0EAwBC"}