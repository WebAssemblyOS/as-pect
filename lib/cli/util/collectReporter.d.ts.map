{"version":3,"file":"collectReporter.d.ts","sourceRoot":"","sources":["../../../src/cli/util/collectReporter.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,YAAY,EAAE,MAAM,yBAAyB,CAAC;AAIvD,OAAO,EAAE,MAAM,EAAE,MAAM,UAAU,CAAC;AAOlC;;;;GAIG;AACH,wBAAgB,eAAe,CAAC,KAAK,EAAE,MAAM,GAAG,YAAY,CAsD3D"}