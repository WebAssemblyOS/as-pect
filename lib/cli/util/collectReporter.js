var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../reporter/DefaultTestReporter", "../../reporter/EmptyReporter", "../../reporter/SummaryTestReporter", "path", "../../reporter/CSVTestReporter", "../../reporter/JSONTestReporter", "querystring", "chalk"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DefaultTestReporter_1 = require("../../reporter/DefaultTestReporter");
    var EmptyReporter_1 = require("../../reporter/EmptyReporter");
    var SummaryTestReporter_1 = require("../../reporter/SummaryTestReporter");
    var path_1 = __importDefault(require("path"));
    var CSVTestReporter_1 = require("../../reporter/CSVTestReporter");
    var JSONTestReporter_1 = require("../../reporter/JSONTestReporter");
    var querystring_1 = __importDefault(require("querystring"));
    var chalk_1 = __importDefault(require("chalk"));
    /**
     * This method inspects the command line arguments and returns the corresponding TestReporter.
     *
     * @param {IYargs} yargs - The command line arguments.
     */
    function collectReporter(yargs) {
        var reporterInput = yargs.argv.reporter || yargs.argv.r;
        var result = /([a-z\.]+)(?:\?(.*))?/i.exec(reporterInput);
        var targetReporter = result ? result[1] : "";
        var options = querystring_1.default.parse(result ? result[2] || "" : "");
        // get relative reporters
        if (targetReporter.startsWith(".")) {
            try {
                var reporterPath = path_1.default.join(process.cwd(), targetReporter);
                var reporterResult = require(reporterPath);
                // if something is returned
                if (reporterResult) {
                    if (typeof reporterResult === "function") { // instantiate it if it's a default exported class
                        return new reporterResult(options);
                    }
                    if (typeof reporterResult.default === "function") {
                        return new reporterResult.default();
                    }
                    else {
                        return reporterResult.default || reporterResult;
                    }
                }
                else {
                    console.log(chalk_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["{bgBlack.yellow [Warning]} Cannot find reporter at {yellow ", "}, defaulting to DefaultTestReporter."], ["{bgBlack.yellow [Warning]} Cannot find reporter at {yellow ", "}, defaulting to DefaultTestReporter."])), reporterPath));
                    return new DefaultTestReporter_1.DefaultTestReporter(options);
                }
            }
            catch (ex) {
                console.log(chalk_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["{bgBlack.yellow [Error]} An error occured while trying to resolve a reporter at {yellow ", "}."], ["{bgBlack.yellow [Error]} An error occured while trying to resolve a reporter at {yellow ", "}."])), targetReporter));
                console.error(ex);
                process.exit(1);
                // @ts-ignore: the process has exited
                return null;
            }
        }
        else if (targetReporter === "EmptyReporter") {
            return new EmptyReporter_1.EmptyReporter(options);
        }
        else if (targetReporter === "SummaryTestReporter") {
            console.log("Hit");
            return new SummaryTestReporter_1.SummaryTestReporter(options);
        }
        else if (targetReporter === "CSVTestReporter") {
            return new CSVTestReporter_1.CSVTestReporter(options);
        }
        else if (targetReporter === "JSONTestReporter") {
            return new JSONTestReporter_1.JSONTestReporter(options);
        }
        else {
            return new DefaultTestReporter_1.DefaultTestReporter(options);
        }
    }
    exports.collectReporter = collectReporter;
    var templateObject_1, templateObject_2;
});
//# sourceMappingURL=collectReporter.js.map