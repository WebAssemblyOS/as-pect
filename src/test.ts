/**
 * This file is used by the package.json "asp" script to simulate using the command line interface.
 */
require("../lib/cli/index.js").asp(process.argv.slice(2));
