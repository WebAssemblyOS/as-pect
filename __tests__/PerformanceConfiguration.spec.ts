import { IPerformanceConfiguration } from "../src/util/IPerformanceConfiguration";
import { TestContext } from "../src/test/TestContext";
import { createPerformanceConfigurationModule } from "./setup/createPerformanceConfigurationModule";

let ctx: TestContext;

let start = new Promise<void>((resolve, reject) => {
  createPerformanceConfigurationModule({}, (err, result) => {
    if (err) {
      console.log(err);
      reject(err);
    } else {
      ctx = result!;
      resolve();
    }
  });
});


describe("Performance Configuration", () => {
  it("should have no warings if maxSamples configured correctly", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      maxSamples: 1000
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    expect(context.warnings).toHaveLength(0);
  });

  it("should have warnings if configured to run less than 0 max samples", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      maxSamples: -1
    };
    const context: TestContext = new TestContext({ performanceConfiguration });

    for (const warning of context.warnings) {
      expect(warning.type).toMatchSnapshot("type");
      expect(warning.message).toMatchSnapshot("message");
    }
  });

  it("should have warnings if configured to run less than allowed max samples", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      maxSamples: 999999999999999
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    for (const warning of context.warnings) {
      expect(warning.type).toMatchSnapshot("type");
      expect(warning.message).toMatchSnapshot("message");
    }
  });

  it("should have warnings if configured maxTestRuntime is too large", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      maxTestRunTime: 999999999999999
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    for (const warning of context.warnings) {
      expect(warning.type).toMatchSnapshot("type");
      expect(warning.message).toMatchSnapshot("message");
    }
  });

  it("should have warnings if configured maxTestRuntime is too negative", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      maxTestRunTime: -1
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    for (const warning of context.warnings) {
      expect(warning.type).toMatchSnapshot("type");
      expect(warning.message).toMatchSnapshot("message");
    }
  });

  it("should have not warnings if configured maxTestRuntime is configured correctly", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      maxTestRunTime: 500
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    expect(context.warnings).toHaveLength(0);
  });

  it("should have warnings if configured decimalPlaces is too large", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      roundDecimalPlaces: 100
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    for (const warning of context.warnings) {
      expect(warning.type).toMatchSnapshot("type");
      expect(warning.message).toMatchSnapshot("message");
    }
  });

  it("should have warnings if configured decimalPlaces is negative", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      roundDecimalPlaces: -1
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    for (const warning of context.warnings) {
      expect(warning.type).toMatchSnapshot("type");
      expect(warning.message).toMatchSnapshot("message");
    }
  });

  it("should have not warnings if configured decimalPlaces is configured correctly", () => {
    const performanceConfiguration: IPerformanceConfiguration = {
      roundDecimalPlaces: 3
    };
    const context: TestContext = new TestContext({ performanceConfiguration });
    expect(context.warnings).toHaveLength(0);
  });

  it("should catch performanceConfiguration warnings at runtime", async () => {
    await start;
    for (const warning of ctx.warnings) {
      expect(warning.type).toMatchSnapshot("type");
      expect(warning.message).toMatchSnapshot("message");
    }
  });
});
