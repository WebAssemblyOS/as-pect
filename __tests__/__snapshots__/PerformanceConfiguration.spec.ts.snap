// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: message 1`] = `"Invalid Performance Configuration: maxSamples exceeds 10000"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: message 2`] = `"Invalid Performance Configuration: maxSamples less than 0."`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: message 3`] = `"Invalid Performance Configuration: maxTestRunTime exceeds 5000"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: message 4`] = `"Invalid Performance Configuration: maxTestRunTime less than 0."`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: message 5`] = `"Invalid Performance Configuration: roundDecimalPlaces exceeds 8"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: message 6`] = `"Invalid Performance Configuration: roundDecimalPlaces less than 0"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: type 1`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: type 2`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: type 3`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: type 4`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: type 5`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should catch performanceConfiguration warnings at runtime: type 6`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should have warnings if configured decimalPlaces is negative: message 1`] = `"Invalid Performance Configuration: roundDecimalPlaces less than 0"`;

exports[`Performance Configuration should have warnings if configured decimalPlaces is negative: type 1`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should have warnings if configured decimalPlaces is too large: message 1`] = `"Invalid Performance Configuration: roundDecimalPlaces exceeds 8"`;

exports[`Performance Configuration should have warnings if configured decimalPlaces is too large: type 1`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should have warnings if configured maxTestRuntime is too large: message 1`] = `"Invalid Performance Configuration: maxTestRunTime exceeds 5000"`;

exports[`Performance Configuration should have warnings if configured maxTestRuntime is too large: type 1`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should have warnings if configured maxTestRuntime is too negative: message 1`] = `"Invalid Performance Configuration: maxTestRunTime less than 0."`;

exports[`Performance Configuration should have warnings if configured maxTestRuntime is too negative: type 1`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should have warnings if configured to run less than 0 max samples: message 1`] = `"Invalid Performance Configuration: maxSamples less than 0."`;

exports[`Performance Configuration should have warnings if configured to run less than 0 max samples: type 1`] = `"PerformanceConfigurationWarning"`;

exports[`Performance Configuration should have warnings if configured to run less than allowed max samples: message 1`] = `"Invalid Performance Configuration: maxSamples exceeds 10000"`;

exports[`Performance Configuration should have warnings if configured to run less than allowed max samples: type 1`] = `"PerformanceConfigurationWarning"`;
